File: algo_2021/s6.-dfs-comb-perm/90.k_sum_II/90_k_sum_ii.java
public class Solution {
    /**
     * @param A: an integer array
     * @param k: a postive integer <= length(A)
     * @param targer: an integer
     * @return: A list of lists of integer
     */
    public List<List<Integer>> kSumII(int[] A, int k, int target) {
        // write your code here
        Arrays.sort(A);
        List<List<Integer>> res = new ArrayList<>();
        dfs(0, A, k, target, new ArrayList<Integer>(), res);
        return res;
    }
    private void dfs(int index, 
                     int[] A, 
                     int k, 
                     int target, 
                     ArrayList<Integer> subset, 
                     List<List<Integer>> res) {
        if (k == 0 && target == 0) {
            res.add(new ArrayList<Integer>(subset));
            return;
        }
        // pruning
        if (k == 0 || target < 0) {
            return;
        }
        for (int i = index; i < A.length; i++) {
            subset.add(A[i]);
            dfs(i + 1, A, k - 1, target - A[i], subset, res);
            subset.remove(subset.size() - 1);
        }
    }
}

File: algo_2021/s6.-dfs-comb-perm/90.k_sum_II/90_k_sum_ii.py
class Solution:
    """
    @param A: an integer array
    @param k: a postive integer <= length(A)
    @param targer: an integer
    @return: A list of lists of integer
    """
    def kSumII(self, A, k, target):
        # write your code here
        A = sorted(A)
        res = []
        self.dfs(0, A, k, target, [], res)
        return res
    def dfs(self, index, A, k, target, subset, res):
        if k == 0 and target == 0:
            res.append(list(subset))
            return 
        # pruning 
        if k == 0 or target < 0:
            return 
        for i in range(index, len(A)):
            subset.append(A[i])
            # need to i + 1, since don't want to pick the same number 
            self.dfs(i + 1, A, k - 1, target - A[i], subset, res)
            subset.pop()